### DISCUSSION TITLE
dark mode dropdown issues

### DISCUSSION THREAD

**Discussion by lanbau** [csandman/chakra-react-select#135] [Q&A] [answered]
### Description

added this package and was expecting dark mode to work out of the box. But as you can see the options are not visible. Any idea how i can style the options background manually? Can't seem to select it using inspect element

<img width="588" alt="Screenshot 2022-05-23 at 11 18 52 PM" src="https://user-images.githubusercontent.com/5410693/169852336-4a543c18-3a19-4139-aa2a-0f06f1fca2c7.png">

### chakra-react-select Version

4.0.0

### TypeScript?

- [ ] Yes I use TypeScript

### Steps to reproduce

2. Click on '...'
3. Scroll down to '...'

### Operating System

- [ ] macOS
- [ ] Windows
- [ ] Linux
- [ ] iOS/iPadOS
- [ ] Android

**Comment by csandman** [comment]
I tested this out and I'm not seeing anything causing this.  See here for a working example: https://codesandbox.io/s/chakra-react-select-demo-65ohb?file=/package.json

I'd need a reproduction to figure out what's causing this.

**Comment by csandman** [comment]
The only thing I would think might be causing this would be that you have modified colors in your chakra theme. If you have any theme customizations, can you post them?

**Comment by aaronleopold** [comment]
I am also facing this issue, happy to share my usage! I'll link the relevant pieces of information, please let me know if there is anything else you might need. 

Here is my [theme config](https://github.com/aaronleopold/stump/blob/develop/packages/core/frontend/src/util/theme.ts)
This is the [component](https://github.com/aaronleopold/stump/blob/develop/packages/core/frontend/src/components/TagSelect.tsx) I am working on (pushed early so I can showcase what's happening). I set `menuIsOpen` so I can inspect it, @lanbau that should help you out debugging since you said you can't select it to inspect it.

### chakra-react-select Version

4.0.0

### TypeScript?

- [x]  Yes I use TypeScript

### Operating System

- [x]  macOS
- [ ]  Windows
- [ ]  Linux
- [ ]  iOS/iPadOS
- [ ]  Android

<img width="598" alt="Screen Shot 2022-05-27 at 11 30 30 AM" src="https://user-images.githubusercontent.com/36278431/170769987-81b319f6-ef4b-4148-9b8b-937622458581.png">

**Comment by jhintux** [comment]
Hi, I am also facing this issue, I think it might be the css. I've noticed that in the demo you provided @csandman 

The button has `--chakra-colors-whiteAlpha-300`
![image](https://user-images.githubusercontent.com/43045722/170831270-b51397dd-e9c7-4695-ae04-21b09cce539b.png)

This is what I imported in my project, it uses `--chakra-colors-gray-100`
![image](https://user-images.githubusercontent.com/43045722/170831301-01b07755-04ee-49c3-a05c-3dab38a16f61.png)

This is my theme config file, I just enabled dark mode https://gist.github.com/edulanasca/71679919953fcb4cee88506393ab62f4

Hope this can help

**Comment by aaronleopold** [comment]
Another update, a friend of mine tried on a windows machine on a Firefox browser and it was working:

![image](https://user-images.githubusercontent.com/36278431/170840171-247bbf0d-66e3-45a9-ab19-decd766be173.png)

**Comment by lanbau** [comment]
i overcame it by forcing a charkra style... but it's still a hack.. would expect it to be an out of the box behaviour

```
import type { ChakraStylesConfig } from "chakra-react-select";

const chakraStyles: ChakraStylesConfig = {
    dropdownIndicator: (provided, state) => ({
      ...provided,
      background: state.isFocused ? "blue.100" : "002C5C",
      p: 0,
      w: "40px",
    }),
    option: (provided, state) => ({
      ...provided,
      background: "#002C5C",
    }),
  };
```

```
<Select<any>
  {...field}
  chakraStyles={chakraStyles}
   isMulti
   options={options}
   placeholder="Select countries"
   closeMenuOnSelect={false} 
/>
```

**Comment by csandman** [comment]
I can see that multiple people are having this issue so I believe that there is some common thread that's causing it, but for some reason I still can't seem to make the issue happen.

I think the biggest clue is the comment from @edulanasca about how the styling is using `gray.100` instead of `whiteAlpha.300` on his local version.  This confuses me however, specifically because I'm not setting that value manually.  It is coming directly from the theme styles for the `addon` component from the `Input`'s multi-part component theme.

The same goes for the dropdown part, which is using the theme styles from `list` component from the `Menu`'s theme. So if it's happening for this, I'd think it would also be happening for the normal menu component/input addon component? It would be good to know if anyone could confirm that.

Besides that, it would also be very helpful to know what specific package versions of chakra, emotion, etc. that people are on so I can try and reproduce this more easily.  I'll continue trying to reproduce this in the meantime.

**Comment by csandman** [comment]
Also @aaronleopold I just tested your `develop` branch on my machine (M1 MacBook Air, with Chrome) and it appears to be working correctly for me:

<img width="607" alt="image" src="https://user-images.githubusercontent.com/9214195/171267855-797fe68c-b18f-4d92-b045-749cd8bf7588.png">

At this point I feel like this issue may be with a specific version of one of the sub-packages of Chakra, or something OS specific but its hard to tell.  Without a reproduction, I'm not sure there's much I can do, because no matter what I try I can't reproduce this issue myself.

---

On a side note, @aaronleopold stump looks like a really cool project!  I've been getting interested in self hosted comic servers recently, and I'm currently a big fan of Komga, but yours looks like it has potential.

**Comment by aaronleopold** [comment]
> I just tested your develop branch on my machine (M1 MacBook Air, with Chrome) and it appears to be working correctly for me

Super strange! If I come across anything that helps resolve this issue I can update the thread here, otherwise I'm okay keeping the manual styling I've added for now.

Also, thanks!! It's a long ways away until a release, but I'm glad you liked it and that you got it up and running! I was a little worried people would have issues after helping my friend get it running on windows 😅

**Comment by csandman** [comment, marked as answer]
Ah, I've found the issue.  This is caused by using `chakra-react-select@^4.0.0` with a version of Chakra UI below 2 (such as 1.8.8).

- React 18 + Chakra 1.8 + CRS 4: https://codesandbox.io/s/react-18-chakra-1-8-crs-4-example-8q5ge4?file=/index.js
- React 17 + Chakra 1.8 + CRS 4: https://codesandbox.io/s/react-17-chakra-1-8-crs-4-example-z0gksz?file=/index.js

As of `chakra-react-select` v4, you must be on `@chakra-ui/react@^2.0.0` as well as `react@^18.0.0` in order for it to work properly, as mentioned in the [release notes](https://github.com/csandman/chakra-react-select/releases/tag/v4.0.0) and the [readme](https://github.com/csandman/chakra-react-select#readme).  You can check Chakra UI's [official migration guide](https://chakra-ui.com/guides/migration) for getting set up with the new version.

If you do not plan on updating your react and chakra versions (which should be done together anyway, Chakra UI v2 is only fully compatible with React v18), please downgrade `chakra-react-select` to v3. I will be maintaining v3 for the foreseeable future to account for this divide, and the only difference currently is the dependent package versions.  If you do stay on v3, I have written [separate docs for it on the v3 branch](https://github.com/csandman/chakra-react-select/tree/v3).

**Reply by csandman** [reply]
After some further evaluation prompted by [this comment](https://github.com/csandman/chakra-react-select/discussions/135#discussioncomment-2874260) I realized that the issue is more generally caused by duplicate versions of `@chakra-ui/system` being installed.  To prevent this, `@chakra-ui/system` has been moved from `dependencies` to `peerDependencies`.  See #136 for more info.

This change has been published in [`v4.0.1`](https://github.com/csandman/chakra-react-select/releases/tag/v4.0.1).

**Comment by aaronleopold** [comment]
I had those versions correct, so just deleted `node_modules` and reinstalled everything and it worked. Must have had residuals from old versions, which is why I'm assuming when you tried my repo is worked out the gate for you. Thanks @csandman!

**Comment by csandman** [comment]
Good to know!  If anyone else is running into this issue after updating everything, try doing a clean install.

**Comment by jhintux** [comment]
Yup, it's working now, I forgot to mention I'm on Windows, with Chrome. Thanks @csandman

**Comment by csandman** [comment]
@edulanasca glad to hear it!

And @aaronleopold I realized that this might be an issue because I noticed in your `package.json` in your main branch you had these dependency versions:

```json
"@chakra-ui/react": "^1.8.8",
"chakra-react-select": "^4.0.0",
"react": "^18.1.0",
"react-dom": "^18.1.0",
```

So there's a fair chance you ended up with your main dependencies while on your development branch.

**Comment by csandman** [comment]
I'm going to convert this to a discussion so I can highlight the answer

**Comment by frysztak** [comment]
this also happens for React 18, Chakra React Select 4 and Chakra UI 2.1. it's Chakra UI 2.1 that causes problems. 

I think it happens because `@chakra-ui/*` packages are listed as dependencies with versions `^2.0.0`. this caused Yarn 3 to resolve `@chakra-ui/system` to version `2.0.2`, which meant that I had two different versions (`2.0.2` and `2.1.1`) running at the same time. this is obviously not good.

temporary workaround for me is to force to use 2.1 version: `yarn set resolution -s @chakra-ui/system@npm:^2.0.0 ^2.1.1` (this updates `yarn.lock` so it's persistent). I believe that moving `@chakra-ui/*` packages to `peerDeps` is the way to go

**Reply by csandman** [reply]
Afraid to say I can't move them to `peerDeps`, I recently had to move them to normal dependencies from dev dependencies to solve a whole other mess of other issues.  However, `chakra-react-select`'s dependencies for chakra's subpackages being pinned at `^2.0.0` shouldn't be an issue as that should allow all versions from `>=2` to `<3`.  If you run into an issue where for some reason CRS had it's dependencies pinned at a lower version it should easily be fixed by running `yarn dedupe` or `yarn dedupe --strategy highest`.

**Reply by csandman** [reply]
However, I'll think on this more, I have to learn more about how `yarn` and `npm` lockfiles handle diverging dependencies among child packages.  If this ends up being an issue every time `@chakra-ui/system` is updated then I'll reconsider changing the way things are organized.  Unfortunately, there seems to be no easy answer for how to handle these deps vs peer deps that works consistently with npm, yarn v1, and yarn v2/v3.  Something always seems to cause a problem.

**Reply by frysztak** [reply]
> dependencies for chakra's subpackages being pinned at ^2.0.0 shouldn't be an issue as that should allow all versions from >=2 to <3

I know, right? yet Yarn was horribly stubborn :laughing: 

> yarn dedupe or yarn dedupe --strategy highest

this is great, I didn't know about this command. thanks!

**Reply by csandman** [reply]
Well... I gave in and went back to peer dependencies for all the packages.  Switching only the `@chakra-ui/system` package to peer deps ended up breaking the rest of the chakra packages I was using on all but one version of `@chakra-ui/react`.  Trying to set up these dependencies in a way that doesn't cause issues for _anyone_ has been the biggest pain point in this project haha.

**Reply by frysztak** [reply]
hopefully it works out :crossed_fingers:

**Comment by sholzmayer** [comment]
For me the Options were shown in darkmode, while the application was in lightmode. The reason for that: The property `value` was missing on the Option item. This library doesn't have a typecheck if the value is there, which would avoid this kind of issues.

**Reply by csandman** [reply]
This sounds like a different issue as the issue mentioned here had to do with duplicate versions of `@chakra-ui/system`.  Do you have an example you could show me (or could you make one with [the CodeSandbox template](https://codesandbox.io/s/chakra-react-select-typescript-4sce1?file=/app.tsx))?  If this is a real issue its something I'd like to investigate.

And as far as type checking goes, there should be type checking if you pass in the structure of the `options` object as a generic to the component.  Something like this:

```tsx
import { Select, OptionBase } from "chakra-react-select";

interface ColorOption extends OptionBase {
  label: string;
  value: string;
}

export const colorOptions = [
  { value: "blue", label: "Blue" },
  { value: "purple", label: "Purple" },
  { value: "red", label: "Red" },
  { value: "orange", label: "Orange" },
  { value: "yellow", label: "Yellow" },
  { value: "green", label: "Green" }
];

const App = () => (
  <Select<ColorOption>
    options={colorOptions}
  />
);
```

If you pass a generic with `label` and `value` and pass in an array of options that don't have the same shape, you'll get type errors.

Here's an example: https://codesandbox.io/s/chakra-react-select-option-generic-example-guiqnw

And in this example, I'm not seeing any dark mode so I'm really curious what you were experiencing.

**Reply by sholzmayer** [reply]
Having status instead of value in a non multi select. See here: https://codesandbox.io/s/chakra-react-select-option-generic-example-forked-stosf1?file=/app.tsx:349-467

Indeed typescript catches this but it breaks the select.

```js
{ status: "orange", label: "Orange" },
{ status: "yellow", label: "Yellow" },
{ status: "green", label: "Green" }
```

![select-status](https://user-images.githubusercontent.com/20296626/194550405-d2a478c3-fbb0-4ea8-9c47-43ae4396c326.gif)

**Reply by csandman** [reply]
I'm not really sure what territory we're entering here haha, this seems like an example that should be breaking.  And in line with that, TypeScript says it's wrong.

If you'd like to use `status` instead of `value` you can however, as long as each item has a unique value for the `status` key (and has it defined).  Just change your `Option` interface to have `status` instead of `value`:

```ts
interface ColorOption extends OptionBase {
  label: string;
  status: string;
}
```

And then pass the prop `getOptionValue` into your Select which returns the `status` key:

```tsx
<Select<ColorOption, true, GroupBase<ColorOption>>
  getOptionValue={({ status }) => status}
/>
```

Here's a complete example:

```tsx
interface ColorOption extends OptionBase {
  label: string;
  status: string;
}

const colorOptions: ColorOption[] = [
  { status: "blue", label: "Blue" },
  { status: "purple", label: "Purple" },
  { status: "red", label: "Red" },
  { status: "orange", label: "Orange" },
  { status: "yellow", label: "Yellow" },
  { status: "green", label: "Green" },
];

const App = () => (
  <Select<ColorOption, true, GroupBase<ColorOption>>
    name="colors"
    options={colorOptions}
    placeholder="Select some colors..."
    getOptionValue={({ status }) => status}
  />
);
```

CodeSandbox: https://codesandbox.io/s/chakra-react-select-option-generic-example-with-getoptionvalue-tc8gz1?file=/app.tsx

---

Side note, when using a single select make sure the second generic on the Select is `false` as that value reflects the `IsMulti` type.
